src/dsl/verifier/proxy/stateHandler/setupStates.ts

```diff
@@ -1,14 +1,29 @@
 import logger from '../../../../common/logger';
-import { ProviderState, ProxyOptions } from '../types';
+import { ProxyOptions, ProviderState, ProviderStateV3 } from '../types';
+
+const isProviderStateV3 = (
+  descriptor: ProviderStateV3 | ProviderState
+): descriptor is ProviderStateV3 =>
+  (descriptor as ProviderStateV3).action !== undefined;
 
 // Lookup the handler based on the description, or get the default handler
 export const setupStates = (
-  descriptor: ProviderState,
+  descriptor: ProviderState | ProviderStateV3,
   config: ProxyOptions
 ): Promise<unknown> => {
   const promises: Array<Promise<unknown>> = [];
 
-  if (descriptor.states) {
+  if (isProviderStateV3(descriptor)) {
+    const handler = config.stateHandlers
+      ? config.stateHandlers[descriptor.state]
+      : null;
+
+    if (handler) {
+      promises.push(handler(descriptor.action === 'setup', descriptor.params));
+    } else {
+      logger.warn(`No state handler found for "${descriptor.state}", ignoring`);
+    }
+  } else if (descriptor && descriptor?.states) {
     descriptor.states.forEach((state) => {
```

Comment:
I know it was like this when you got here, but I think we should do this as a `map` rather than a `forEach` that produces an array.

Reply:
Also, does the V3 verifier sometimes have an array of states? Or does it just fire multiple requests?

Reply:
> I know it was like this when you got here, but I think we should do this as a `map` rather than a `forEach` that produces an array.

Just having a look. I think you're probably going for "immutability". Because there are multiple code paths that may not return a promise, we'll need to guard for nulls etc. I think

I'll see what I can do.

> Also, does the V3 verifier sometimes have an array of states? Or does it just fire multiple requests?

It fires multiple, one per state. Technically, the array is only a v3 thing from Ruby, so the Rust verifier will actually never experience this.

Now that I look at the source, I think it will _always_ receive a `ProviderStateV3` so I can actually cleanup this code anyway.